eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/ZmqWatcher.scala

```diff
@@ -226,13 +238,128 @@ class ZmqWatcher(chainHash: ByteVector32, blockCount: AtomicLong, client: Extend
+            log.warning("not enough funds to publish tx, will retry next block: reason={} input={}:{} txid={}", t.getMessage, signingKit.spentOutpoint.txid, signingKit.spentOutpoint.index, p.tx.txid)
+            self ! PublishNextBlock(p)
+            Future.failed(t)
+          case t: Throwable =>
+            log.error("cannot publish tx: reason={} input={}:{} txid={}", t.getMessage, signingKit.spentOutpoint.txid, signingKit.spentOutpoint.index, p.tx.txid)
+            Future.failed(t)
+        }
+      case PublishStrategy.JustPublish =>
+        log.info("publishing tx: txid={} tx={}", p.tx.txid, p.tx)
+        publish(p.tx, isRetry = false)
+    }
+  }
+
+  def publish(tx: Transaction, isRetry: Boolean): Future[ByteVector32] = {
+    client.publishTransaction(tx)(singleThreadExecutionContext).recoverWith {
       case t: Throwable if t.getMessage.contains("(code: -25)") && !isRetry => // we retry only once
         import akka.pattern.after
-        after(3 seconds, context.system.scheduler)(Future.successful({})).map(_ => publish(tx, isRetry = true))
-      case t: Throwable => log.error("cannot publish tx: reason={} txid={} tx={}", t.getMessage, tx.txid, tx)
+        after(3 seconds, context.system.scheduler)(Future.successful({})).flatMap(_ => publish(tx, isRetry = true))
+      case t: Throwable =>
+        log.error("cannot publish tx: reason={} txid={}", t.getMessage, tx.txid)
+        Future.failed(t)
+    }
+  }
+
+  def publishCommitWithAnchor(commitTx: Transaction, currentFeerate: FeeratePerKw, targetFeerate: FeeratePerKw, dustLimit: Satoshi, signingKit: TransactionSigningKit.ClaimAnchorOutputSigningKit): Future[ByteVector32] = {
+    import signingKit._
+    if (targetFeerate <= currentFeerate) {
+      log.info(s"publishing commit tx without the anchor (current feerate=$currentFeerate): txid=${commitTx.txid}")
+      publish(commitTx, isRetry = false)
+    } else {
+      log.info(s"publishing commit tx with the anchor (target feerate=$targetFeerate): txid=${commitTx.txid}")
+      // We want the feerate of the package (commit tx + tx spending anchor) to equal targetFeerate.
+      // Thus we have: anchorFeerate = targetFeerate + (weight-commit-tx / weight-anchor-tx) * (targetFeerate - commitTxFeerate)
+      // If we use the smallest weight possible for the anchor tx, the feerate we use will thus be greater than what we want,
+      // and we can adjust it afterwards by raising the change output amount.
+      val anchorFeerate = targetFeerate + FeeratePerKw(targetFeerate.feerate - currentFeerate.feerate) * commitTx.weight() / Transactions.claimAnchorOutputMinWeight
+      // NB: bitcoind requires txs to have at least one output, but we'll remove it later to keep a single change output.
+      // In case we have the perfect set of utxo amounts and no change output is added, we need the amount to be greater
+      // than the fee because we may need to deduce the fee from that output.
```

Comment:
```suggestion
      // NB: bitcoind requires txs to have at least one output when calling fundrawtransaction, and may add an additional change output. But since the purpose of this transaction is just to do a CPFP, it should really have either no output, or one single change output. To workaround this limitation, we use a dummy output and later merge the dummy output with the optional change output.
      // In case we have the perfect set of utxo amounts and no change output is added, we need the amount to be greater
      // than the fee because we may need to deduce the fee from that output.
```

I don't understand that last phrase:
> In case we have the perfect set of utxo amounts and no change output is added, we need the amount to be greater than the fee because we may need to deduce the fee from that output.

NB: the "no change" scenario is going to happen more often than not, in my experience bitcoin core is pretty smart with utxos selection in that regard

Reply:
You're right, I should clarify this comment. This is a tricky one though...

What this means is that in the case where `bitcoind` doesn't add a change output, we may need to tweak the amount of the dummy change output we added ourselves to reach our desired feerate if the final weight is greater than the weight we estimated.

In that case we will need to *decrease* this output's amount to reach the desired feerate, so we need this amount to initially be high enough to allow that. I *think* the current value should always work out (or worst case we will pay a bit less fees than we'd like but the difference shouldn't be game-changing).

Reply:
I'm not *sure* we need this, but I feel more confident with it :)

Reply:
> if the final weight is greater than the weight we estimated

This would be due to the added input correct?

Reply:
My previous explanation was still a bit unclear (because my memory of it was still unclear), here is what happens:

- we ask bitcoind to create a funded transaction, but with our non-wallet input removed
- bitcoind will add inputs and will take into account the weight of the inputs it adds, but cannot take into account the weight of our non-wallet input (because we removed it since fundrawtransaction doesn't support non-wallet inputs)
- when we re-add our non-wallet input afterwards, we've increased the weight so we've decreased the fee
- we need to lower the amount out to correct that
- if bitcoind didn't add any change output, we need our initial dummy change output to have enough sats that we can "borrow" to pay the fee

Is that clearer? I'll add this to the comment.

Reply:
The very last bullet point is still a bit unclear. Why don't we just always take from the dummy output to increase the tx weight (when we need to), independently of whether there is a change output or not?

Reply:
We do, we always subtract from the total amount out (later in `adjustAnchorOutputChange`).

But we must ensure that the total amount out is high enough, so since we can't control the amount of the change output added by bitcoind, we must ensure that minimal amount is met in our dummy change output. Does that make sense?

Reply:
But why do you mention the bitcoind-generated change output in that bullet point at all? Why not just:

- we need our initial dummy change output to have enough sats that we can take from it to pay the fee

Reply:
I get it, the missing part is you assume that you have already merged the bitcoind-generated change output and the dummy change output.

Reply:
Yes, here is how I'm updating the comment, let me know if that's more clear:

```
      // NB: fundrawtransaction requires at least one output, and may add at most one additional change output.
      // Since the purpose of this transaction is just to do a CPFP, the resulting tx should have a single change output
      // (note that bitcoind doesn't let us publish a transaction with no outputs).
      // To work around these limitations, we start with a dummy output and later merge that dummy output with the optional
      // change output added by bitcoind.
      // NB: fundrawtransaction doesn't support non-wallet inputs, so we have to remove our anchor input and re-add it later.
      // That means bitcoind will not take our anchor input's weight into account when adding inputs to set the fee.
      // That's ok, we can increase the fee later by decreasing the output amount. But we need to ensure we'll have enough
      // to cover the weight of our anchor input, which is why we set it to the following value.
```

Reply:
Perfect!

Reply:
Done in https://github.com/ACINQ/eclair/pull/1702/commits/090f7cba3b80eb4335eab9716b7b2154db5c4ef3